From a2aae9593aa76e4ba05c52892bb6e49fcfabb34d Mon Sep 17 00:00:00 2001
From: Ricardo Cerqueira <cyanogenmod@cerqueira.org>
Date: Mon, 4 Nov 2013 02:54:13 +0000
Subject: [PATCH] wifi: Set the mode at native layer

 * Set normal/AP mode so that the HAL can specify correct module
   arguments when loading the driver.

Change-Id: I6eee2cc53b5f1e70306e252e52c909833f220515
---
 service/java/com/android/server/wifi/WifiNative.java       | 2 ++
 service/java/com/android/server/wifi/WifiStateMachine.java | 2 ++
 service/jni/com_android_server_wifi_WifiNative.cpp         | 6 ++++++
 3 files changed, 10 insertions(+)

diff --git a/service/java/com/android/server/wifi/WifiNative.java b/service/java/com/android/server/wifi/WifiNative.java
index 07f3abf..4d42bff 100644
--- a/service/java/com/android/server/wifi/WifiNative.java
+++ b/service/java/com/android/server/wifi/WifiNative.java
@@ -1547,6 +1547,8 @@ public class WifiNative {
         }
     }
 
+    public native static boolean setMode(int mode);
+
     /* WIFI HAL support */
 
     // HAL command ids
diff --git a/service/java/com/android/server/wifi/WifiStateMachine.java b/service/java/com/android/server/wifi/WifiStateMachine.java
index 6fff53a..5f0c327 100644
--- a/service/java/com/android/server/wifi/WifiStateMachine.java
+++ b/service/java/com/android/server/wifi/WifiStateMachine.java
@@ -1785,6 +1785,7 @@ public class WifiStateMachine extends StateMachine implements WifiNative.WifiRss
      */
     public void setSupplicantRunning(boolean enable) {
         if (enable) {
+            WifiNative.setMode(0);
             sendMessage(CMD_START_SUPPLICANT);
         } else {
             mDisconnectDelayDuration = -1;
@@ -1812,6 +1813,7 @@ public class WifiStateMachine extends StateMachine implements WifiNative.WifiRss
      */
     public void setHostApRunning(WifiConfiguration wifiConfig, boolean enable) {
         if (enable) {
+            WifiNative.setMode(1);
             sendMessage(CMD_START_AP, wifiConfig);
         } else {
             sendMessage(CMD_STOP_AP);
diff --git a/service/jni/com_android_server_wifi_WifiNative.cpp b/service/jni/com_android_server_wifi_WifiNative.cpp
index 80a19c0..3ce7bf0 100644
--- a/service/jni/com_android_server_wifi_WifiNative.cpp
+++ b/service/jni/com_android_server_wifi_WifiNative.cpp
@@ -136,6 +136,11 @@ static jstring doStringCommand(JNIEnv* env, jstring javaCommand) {
     return env->NewString((const jchar *)str.string(), str.size());
 }
 
+static jboolean android_net_wifi_setMode(JNIEnv* env, jobject, jint type)
+{
+    return (jboolean)(::wifi_set_mode(type) == 0);
+}
+
 static jboolean android_net_wifi_isDriverLoaded(JNIEnv* env, jclass)
 {
     return (::is_wifi_driver_loaded() == 1);
@@ -2605,6 +2610,7 @@ static JNINativeMethod gWifiMethods[] = {
     { "doIntCommandNative", "(Ljava/lang/String;)I", (void*)android_net_wifi_doIntCommand },
     { "doStringCommandNative", "(Ljava/lang/String;)Ljava/lang/String;",
             (void*) android_net_wifi_doStringCommand },
+    { "setMode", "(I)Z", (void*) android_net_wifi_setMode},
     { "startHalNative", "()Z", (void*) android_net_wifi_startHal },
     { "stopHalNative", "()V", (void*) android_net_wifi_stopHal },
     { "waitForHalEventNative", "()V", (void*) android_net_wifi_waitForHalEvents },
-- 
2.9.3

